### ISSUE TITLE
DevTools: Improve browser extension iframe support

### ISSUE THREAD

**Issue by dmail** [facebook/react#18945]
When react is inside an iframe, chrome extension for react devtools fails to detect react.

This is because the extension sets `__REACT_DEVTOOLS_GLOBAL_HOOK__` only on the top level window. Apparently it's possible to have `__REACT_DEVTOOLS_GLOBAL_HOOK__` on iframes too by adding `all_frames: true` in extension [manifest.json](https://github.com/facebook/react/blob/c390ab3643612dc08ca4bebadc5b0377e9e7eb79/packages/react-devtools-extensions/chrome/manifest.json#L49). It was done by redux devtools extension in https://github.com/zalmoxisus/redux-devtools-extension/pull/56.

Have you considered adding `all_frames: true` to chrome extension ?

## Steps To Reproduce

1. Create a file `react-main.html`
```html
<!DOCTYPE html>
<html>
  <head>
    <script crossorigin src="https://unpkg.com/react@16/umd/react.development.js"></script>
    <script crossorigin src="https://unpkg.com/react-dom@16/umd/react-dom.development.js"></script>
  </head>
  <body></body>
</html>

```
2. Create a file `react-iframe.html`
```html
<!DOCTYPE html>
<html>
  <head></head>
  <body>
    <iframe src="./react-main.html" />
  </body>
</html>
```
3. Open `react-iframe.html` in chrome

## The current behavior

![react-devtools-not-detected](https://user-images.githubusercontent.com/443639/82196553-21879100-98fa-11ea-932b-b9f49a4079c7.png)

## The expected behavior

![react-devtools-detected](https://user-images.githubusercontent.com/443639/82196588-2cdabc80-98fa-11ea-925d-186e8c999f44.png)

Pull request: https://github.com/facebook/react/pull/18952

**Comment by bvaughn** [comment]
The frame limitation is mentioned in the docs, FWIW:
https://github.com/facebook/react/tree/master/packages/react-devtools

It's not a bug, just a known limitation. We currently suggest using the standalone version (as linked above) or (if you also control the iframe) passing the hook through from the parent:
https://github.com/facebook/react/blob/d897c35ecfb7471b55e41933fd4d993b98dbf600/fixtures/devtools/regression/14.9.html#L9-L11

I don't think that adding the Manifest setting you mentioned would be enough on its own to make the React DevTools extension detect and support versions of React running inside of iframes.

I'd be happy for you to take a shot at implementing it though if you'd like. We have a [regression test](https://github.com/facebook/react/blob/d897c35ecfb7471b55e41933fd4d993b98dbf600/fixtures/devtools/regression/server.js) that you could use to verify this:
```sh
./fixtures/devtools/regression/server.js
# Open localhost:3000 and see if DevTools detects React
```

**Comment by dmail** [comment]
Thank you very much for your answer and rewording the issue title, it's more accurate.

In my case the iframe runs in an other domain so window.parent would not be accessible.

I am ready to try to implement it. I'll take a look tomorrow. If you have any more info to share that would greatly help me when I'll start working on it.

Thanks again, I will keep you informed and I hope I will be able to make it work :)

**Comment by bvaughn** [comment]
Great! Keep me posted on your progress :smile:

Here's instructions to get you started:
https://github.com/facebook/react/tree/master/packages/react-devtools-extensions#build-steps

**Comment by dmail** [comment]
I have opened a draft pull request. 
For now I have only tested to add all_frames: true on chrome extension manifest.json and it looks like it's working -> react is being detected inside a sandboxed iframe 🎉 .

**Comment by dmail** [comment]
Working on edge 🎉 

About firefox however I have some trouble. I have changed nothing yet concerning firefox and after doing

1. yarn build:firefox
2. yarn run test:firefox

In the firefox browser that is launched the extension does not detect react on `reactjs.org`.

![firefoxko](https://user-images.githubusercontent.com/443639/82307985-6fb39780-99c1-11ea-9e0c-f9e3e07969ea.png)

In Firefox console I see this error (and many more with the same error code):

![image](https://user-images.githubusercontent.com/443639/82307376-92917c00-99c0-11ea-8a7f-3d3fb8f47f0c.png)

The extension in firefox store is working properly

![firefoxok](https://user-images.githubusercontent.com/443639/82307857-3ed36280-99c1-11ea-90ec-068114f33ad8.png)

Firefox fails to load some file (apparently related to Localization) when the extension is loaded locally. Any idea what is going on ?

**Comment by dmail** [comment]
About firefox I got workarounds depending what happens:

(a) Firefox addon is shown but fails to detect react
-> refresh the page
-> works normally

(b) Firefox addon does not even show up
-> go to Firefox settings
-> disable react devtools addon and enable it right away
-> go back to `reactjs.org`
-> works normally

As I said this happens already on `master` branch without any change on my side.

Now that I can test firefox extension locally I move on and try to add `all_frame: true` on Firefox.

**Comment by dmail** [comment]
I'm putting pull request in ready for review.

**Comment by shakhbulatgaz** [comment]
What happened to this issue? Is it available for me to work on?
@dmail , are you planning on working on it or I may take it?

**Comment by dmail** [comment]
Hello, the pull request is on hold because I don't know what to tackle first. On my side I was waiting for advice to restart working on it. Go ahead, I'm still interested on the subject so I'll keep an eye on it (and may still help to the extent of my abilities).

**Comment by markerikson** [comment]
For the record, I was able to make this work with a Next.js app that is providing the iframe, with these changes:

**$NEXT_PROJECT/public/js/enableReactDevtoolsIframe**
```js
// The React DevTools do not normally look inside of iframes, just the outer window.
// We can enable inspection of React components in an iframe by copying this global variable:
// https://github.com/facebook/react/issues/18945#issuecomment-630421386
// This code must be injected before React runs, so we add it as a separate tag
// and show it via Next.js <Head> in _app.tsx.
if (window.parent) {
  window.__REACT_DEVTOOLS_GLOBAL_HOOK__ = window.parent.__REACT_DEVTOOLS_GLOBAL_HOOK__;
}
```

**$NEXT_PROJECT/pages/_app.tsx**
```ts
import React from 'react';
import Head from 'next/head';

export default function MyApp({ Component, pageProps }: { Component: any; pageProps: any }) {
  // Ensure the React DevTools global variable is injected if this is an iframe
  // to enable inspection of components inside the iframe.
  return (
    <React.Fragment>
      <Head>
        <script src="js/enableReactDevtoolsIframe.js"></script>
      </Head>
      <Component {...pageProps} />
    </React.Fragment>
  );
}
```

**Comment by marvinhagemeister** [comment]
FYI: We've been using the `all_frames: true` option in `manifest.json` for a while now over with preact-devtools and it works like a charm. The only thing one needs to be careful about is to inject the highlighting code into each iframe too! Otherwise  the position will be off.

**Comment by bvaughn** [comment]
@marvinhagemeister Do you remember doing anything additional (beyond just adding `all_frames: true` option in `manifest.json`) to get it to work? I haven't taken the time to dig into this yet but I did test out the linked PR (#18952) and it didn't appear to be working.

@shakhbulatgaz  did you ever end up digging into this?

**Comment by marvinhagemeister** [comment]
@bvaughn Not much really, only had to ensure that each adapter that's injected into the client doesn't traverse into other iframes. The linked PR looks correct to me, so I'm guessing something isn't ok with the connection logic. Maybe it's the same timing issue we've been talking about a while back that's more of a regression in Chrome?

For reference: The PR that added support for iframes for `preact-devtools`: https://github.com/preactjs/preact-devtools/pull/209/files

**Comment by bvaughn** [comment]
Gotcha! Thanks for elaborating :smile:

> Maybe it's the same timing issue we've been talking about a while back that's more of a regression in Chrome?

I think the fix for this issue has made its way into stable (or at least I'm no longer seeing the issue myself).

**Comment by shakhbulatgaz** [comment]
@bvaughn Nah, I'm sorry, other things came up and I completely forgot about this 😢

**Comment by bvaughn** [comment]
Fair enough. Maybe someone else will pick it up :)

**Comment by NarendraHegade5** [comment]
@bvaughn Can you explain to me the current scenario and what needs to be done? Was reading through the thread but couldn't deduce much. This is my first issue to pick from react repo.

**Comment by anrao91** [comment]
@bvaughn Can you explain to me the current scenario and what needs to be done? Was reading through the thread but couldn't deduce much. This is my first issue to pick from react repo.

**Comment by bvaughn** [comment]
@anrao91 Unfortunately, no. Not beyond what's been discussed on this thread and on the linked PR (#18952).

**Comment by Reflex-Gravity** [comment]
@bvaughn By adding the below code, doesn't enable react to inspect inside an iframe. I tried the [regression test](https://github.com/facebook/react/blob/d897c35ecfb7471b55e41933fd4d993b98dbf600/fixtures/devtools/regression/server.js), didn't work there either.
```javascript
__REACT_DEVTOOLS_GLOBAL_HOOK__ = parent.__REACT_DEVTOOLS_GLOBAL_HOOK__; 
```

I was able to make this work by copying the values of parents `__REACT_DEVTOOLS_GLOBAL_HOOK__` one by one, instead of copying it directly. (Maybe this isn't the right approach, but it seems to work)
```javascript
for (const key in __REACT_DEVTOOLS_GLOBAL_HOOK__) {
      __REACT_DEVTOOLS_GLOBAL_HOOK__[key] = parent.__REACT_DEVTOOLS_GLOBAL_HOOK__[key]
}
```

I verified it by doing the [regression test](https://github.com/facebook/react/blob/d897c35ecfb7471b55e41933fd4d993b98dbf600/fixtures/devtools/regression/server.js) on Firefox and Chrome. It detected React and also displayed the entire component tree under the `Components Panel`. (Didn't work for  **React 0.14.9** though due to incompatibility)

Copying it directly didn't update the iframe's `__REACT_DEVTOOLS_GLOBAL_HOOK__` value, I'm not really sure why though. Maybe because it's is not `writable`. When using `defineProperty`, `writable` is `false` by default. But, In this case, the above working code shouldn't have worked. :smile: 
https://github.com/facebook/react/blob/9ea0f6752da28a91bdb56382367e2a07390cb733/packages/react-devtools-shared/src/hook.js#L309-L321

Below is a sceenshot showing the working eg:,

![react_dev_tools_issue](https://user-images.githubusercontent.com/19655674/87353756-0db77e80-c57b-11ea-8cf8-3c18a3dc105c.png)

**Comment by sarathps93** [comment]
Can I take this up if no one else is currently working on it? @bvaughn @dmail

**Comment by bvaughn** [comment]
@sarathps93 You're welcome to work on this. Please check out the work and discussion on #19345 before starting though!

**Comment by omarsy** [comment]
@sarathps93  You’re still working on it ?

**Comment by sarathps93** [comment]
@omarsy I was about to start this weekend. Let me know if you want to take it up

**Comment by omarsy** [comment]
Can I push a PR ?

**Comment by sarathps93** [comment]
@omarsy Yes please go ahead

**Comment by omarsy** [comment]
Thank you @sarathps93

**Comment by bvaughn** [comment]
Resolved in #19827

**Comment by bvaughn** [comment]
Re-opening because this was reverted again.

**Comment by Eric0926** [comment]
@bvaughn I have walked through previous discussions. Can I take a look at it as my first issue?

**Comment by bvaughn** [comment]
Feel free, @Eric0926.

I'm a little skeptical about landing this change again though, given we've had to revert it twice. 😅 But if you're willing to do a lot of testing, sure!

**Comment by alleksei37** [comment]
It looks like the issue is still relevant. 🙁

**Comment by bvaughn** [comment]
That's why the issue is still open, @alleksei37.

**Comment by KutnerUri** [comment]
can you merge the fix with a feature flag? (i.e. enable-able from settings)
I'm having the `__REACT_DEVTOOLS_GLOBAL_HOOK__` fix and I want to remove it!

**Comment by bvaughn** [comment]
No. If there was a simple fix that just needed to be merged, it would have been merged already. Both previous attempts to fix this have needed to be reverted because they've caused unexpected problems.

**Comment by prsmahajan** [comment]
Hi, I want to work on this issue, can you suggest me ways to get started?

**Comment by bvaughn** [comment]
@prsmahajan You can find an architecture overview doc about how DevTools works here:
https://github.com/facebook/react/blob/master/packages/react-devtools/OVERVIEW.md

And instructions for how to build and test the extension here:
https://github.com/facebook/react/tree/master/packages/react-devtools-extensions#local-development

Beyond that, this would likely be an unguided task because we're all spread a little thin at the moment.

**Comment by prsmahajan** [comment]
What is the main problem you all are facing?

**Comment by flydge** [comment]
Solution above works partly, only with deactivated CORS (**for me**)
Do u have any ideas, how to allow using of devtools with enabled CORS?
![photo_2021-06-19 14 49 30](https://user-images.githubusercontent.com/28603471/122642026-21ffe600-d111-11eb-8987-82f0dfd40044.jpeg)

**Comment by jplindgren** [comment]
I believe I have the same problem as @flydge .
![image](https://user-images.githubusercontent.com/3172397/126051919-b64f2109-8e77-4bdd-8e89-393a4958995c.png)

**Comment by amorphousDj** [comment]
@bvaughn Can I take a look at it as my first issue?

**Comment by bvaughn** [comment]
If you'd like to, please do.

**Comment by luanhcastro** [comment]
Hi i'd like to contribute.

**Comment by richiepreece** [comment]
This seems more urgent with google announcing removing [`document.domain`](https://developer.chrome.com/blog/immutable-document-domain/#alternative-cross-origin-communication). This means the parent hook passthrough is only available for same-protocol/subdomain/domain iframes.

Has there been any movement on this?

**Comment by awells111** [comment]
It looks there are a couple pull requests that try to use the manifest.json approach, setting all_frames: true and getting the global devtools hook from the parent. That hasn't worked for my use case and I'm not confident that I would be able to fix the issue in the repo that way since it looks like that fix has been tried for about 7 years (see [react-devtools issue 76](https://github.com/facebook/react-devtools/issues/76).

As a workaround, what I want to do is create a dropdown that will let the user pick a frame and then show the component tree for that frame. Is that a viable solution? Would I have better luck with using the standalone devtools app found [here](https://github.com/facebook/react/tree/main/packages/react-devtools#usage-with-react-dom)?

**Comment by dmail** [comment]
For the record all_frames: true was inspired from preact-devtools.

**Comment by awells111** [comment]
all_frames: true seems to be a content_scripts setting that tells the scripts listed to run in all frames instead of just the top frame. From [what I saw](https://stackoverflow.com/questions/8917593/do-chrome-extensions-access-iframes), all_frames: true won't help if iframes are dynamically loaded, which doesn't help for my application.

I made a [demo app](https://github.com/awells111/react-iframe-test/) where I tried to get a parent React app to display a child React app in an iframe. The results are pretty consistent with this thread. The parent React app can't see any components inside of the iframe'd child React app.

Setting the child app's __REACT_DEVTOOLS_GLOBAL_HOOK__ to its parent's hook won't work because of the "Uncaught DOMException: Blocked a frame with origin "http://localhost:3002" from accessing a cross-origin frame."

So I want to try to find a solution that will work for dynamically loaded React apps that don't come from the same origin. Right now my idea is adding a "frame selector" to the devtools browser extension but I am open to ideas.

**Comment by awells111** [comment]
@omarsy Do you remember why #19854 was reverted? Is it because the changes cause some tests to fail or is it something else? The changes in that PR look like they're meeting the needs of my application.

**Comment by omarsy** [comment]
Hello the issue is reverted because someone says it has an issue with my PR https://github.com/facebook/react/pull/19854#issuecomment-705190614. But I don't think it is related of my PR.
It is why I create a new PR to fix the issue of the comment https://github.com/facebook/react/pull/19994 to see  if it is caused by my PR or not.
If I have time this month I will try to close my two PRs

**Comment by richiepreece** [comment]
@awells111  I think the frame selector is ideal if possible. Similar to what chrome does when you have iframes.

There's always the option for the standalone dev tools, but honestly it's not great for when you're trying to find a component because you don't have a selector, and you have to know the component tree to know where to look.

**Comment by awells111** [comment]
#19854 fixes the issue for me. I didn't wind up implementing anything like a frame selector.

A fork including the changes from #19854 can't be uploaded to the web store until the manifest is updated to v3, if we're even allowed to upload forks to the chrome web store. As a temporary solution, developers can create their own forks and build the modified repo locally. That has worked for me for now.
